src/WebJobs.Script.WebHost/Metrics/FlexConsumptionMetricsPublisher.cs

```diff
@@ -0,0 +1,301 @@
+
+            var metricsDirectoryInfo = _fileSystem.DirectoryInfo.FromDirectoryName(MetricsFilePath);
+            var files = metricsDirectoryInfo.GetFiles().OrderBy(p => p.CreationTime).ToList();
+
+            // ensure we're under the max file count
+            if (files.Count < _optionsMonitor.CurrentValue.MaxFileCount)
+            {
+                return true;
+            }
+
+            // we're at or over limit
+            // delete enough files that we have space to write a new one
+            int numToDelete = files.Count - _optionsMonitor.CurrentValue.MaxFileCount + 1;
+            var filesToDelete = files.Take(numToDelete).ToArray();
+
+            _logger.LogDebug($"Deleting {filesToDelete.Length} metrics file(s).");
+
+            foreach (var file in filesToDelete)
+            {
+                try
+                {
+                    file.Delete();
+                }
+                catch (Exception ex) when (!ex.IsFatal())
+                {
+                    // best effort
+                    _logger.LogError(ex, $"Error deleting metrics file '{file.FullName}'.");
+                }
+            }
+
+            files = metricsDirectoryInfo.GetFiles().OrderBy(p => p.CreationTime).ToList();
+
+            // return true if we have space for a new file
+            return files.Count < _optionsMonitor.CurrentValue.MaxFileCount;
+        }
+
+        private void OnStandbyOptionsChange()
+        {
+            if (!_standbyOptions.CurrentValue.InStandbyMode)
+            {
+                Start();
+            }
+        }
+
+        public void OnFunctionStarted(string functionName, string invocationId)
+        {
+            if (!_started)
+            {
+                return;
+            }
+
+            lock (_lock)
+            {
+                if (ActiveFunctionCount == 0)
+                {
+                    // we're transitioning from inactive to active
+                    _instanceActivityStopwatch = ValueStopwatch.StartNew();
+                }
+
+                ActiveFunctionCount++;
+            }
+        }
+
+        public void OnFunctionCompleted(string functionName, string invocationId)
+        {
+            if (!_started)
+            {
+                return;
+            }
+
+            lock (_lock)
+            {
+                if (ActiveFunctionCount > 0)
+                {
+                    ActiveFunctionCount--;
+                }
+
+                if (ActiveFunctionCount == 0)
+                {
+                    // we're transitioning from active to inactive accumulate the elapsed time,
+                    // applying the minimum interval
+                    var elapsedMS = _instanceActivityStopwatch.GetElapsedTime().TotalMilliseconds;
+                    var duration = Math.Max(elapsedMS, _optionsMonitor.CurrentValue.MinimumActivityIntervalMS);
+                    FunctionExecutionTimeMS += (long)duration;
```

Comment:
>FunctionExecutionTimeMS [](http://example.com/codeflow?start=20&length=23)

we use += here since within the same 5 second window we might have multiple transitions from active to inactive to active to inactive?

Reply:
Correct. Consider a series of non-overlapped invocations. We'll be transitioning multiple times active/inactive.